src/protocol/boolean/bit_decomposition.rs

```diff
@@ -38,49 +40,170 @@ impl BitDecomposition {
+                (true, false) => G::Q,
+                // Case where f_i - c_i = -1
+                (false, true) => G::NotQ,
+                // this is what g_i should actually be here, so we can use this down the road
+                // let g_i_foo = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+            };
+            g_bin.push(g_i);
+            // let f_i_minus_c_i: u64 = f_i - c_i;
+            // let g_prime: S = q_p.clone().into() * f_i_minus_c_i;
+            // // c_i ∈ {0, 1}, so F::truncate_from will work
+            // debug_assert!(c_i <= 1_u128);
+            // g_bin.push(S::share_known_value(&ctx, F::truncate_from(c_i)) + &g_prime);
+        }
+
+        // Step 5. Compute BitwiseSum([r]_B, [g]_B)
+        let mut h: Vec<S> = Vec::with_capacity(usize::try_from(el).unwrap());
+        let one_minus_q_p = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+        let (mut last_carry, result_bit, mut last_carry_known_to_be_zero) = match g_bin[0] {
+            G::Zero => (S::ZERO, r.b_b[0].clone(), true),
+            G::One => (r.b_b[0].clone(), S::share_known_value(&ctx, F::ONE), false),
+            G::Q => (
+                q_p.multiply(&r.b_b[0], ctx.narrow(&Step::AddGtoR), record_id)
+                    .await?,
+                q_p.clone() + &r.b_b[0],
+                false,
+            ),
+            G::NotQ => (
+                one_minus_q_p
+                    .clone()
+                    .multiply(&r.b_b[0], ctx.narrow(&Step::AddGtoR), record_id)
+                    .await?,
+                one_minus_q_p.clone() + &r.b_b[0],
+                false,
+            ),
+        };
+        h.push(result_bit);
 
-        // Step 8, 9. [g_i] = [q] * f_i
-        // Step 10. [h]_B = [d + g]_B, where [h]_B = ([h]_0,...[h]_(el+1))
-        // Step 11. [a]_B = ([h]_0,...[h]_(el-1))
-        let a_b =
-            maybe_add_constant_mod2l(ctx.narrow(&Step::AddDtoG), record_id, &d_b, x, &q_p).await?;
+        for (bit_index, bit) in r.b_b.iter().enumerate().skip(1) {
+            let mult_result = if last_carry_known_to_be_zero {
+                // TODO: this makes me sad
+                S::ZERO
+                    .multiply(&S::ZERO, ctx.narrow(&BitOpStep::from(bit_index)), record_id) // this is stupid
+                    .await?;
 
-        Ok(a_b)
+                S::ZERO
+            } else {
+                last_carry
+                    .multiply(bit, ctx.narrow(&BitOpStep::from(bit_index)), record_id)
+                    .await?
+            };
+            let last_carry_or_bit = -mult_result.clone() + &last_carry + bit;
+            let next_carry = match &g_bin[bit_index] {
+                G::Zero => mult_result,
+                G::One => {
+                    last_carry_known_to_be_zero = false;
+                    last_carry_or_bit
+                }
+                G::Q => {
+                    last_carry_known_to_be_zero = false;
+                    q_p.multiply(
```

Comment:
I am not sure this will work. Values seem to be chosen by rolling a dice, so for record = 1 you may hit this path while record 0 may short-circuit it if bit is zero or one. 

There is a limitation that the number of multiplications (or communications) per protocol must be deterministic

Unless I misunderstood this code, it is not the case here

Reply:
oh well that explains what's going on in line 129! that does make me sad.

this probably means we won't get any wins here, since some of the "speed up" would be from knowing this value ~half the time, randomly